### DISCUSSION TITLE
Internal Server Error when useState initializer function uses localStorage in custom hook.

### DISCUSSION THREAD

**Discussion by kaganAhmetOkan** [vercel/next.js#61450] [Help]
### Summary

Using `localStorage` in an initializer function of a useState hook in my custom hook causes `Internal Server Error 500` on client side, even though everything works perfectly. The problem is that Next.js is trying to call the function from the server, even though it's a custom hook.

```ts
// inside custom hook
function initializeState() {
  const oldState = localStorage.getItem(key);
  if (oldState !== null) return JSON.parse(oldState);
  return initialState;
};

const [state, setState] = useState(initializeState);
```

Both `key` and `initialState` are props passed to the custom hook. In this example, calling `localStorage` should be perfectly fine since this is a custom hook, but Next.js runs this function from the server which causes an internal server error.

Adding `if (typeof window === "undefined" return initialState;` as a precaution causes `prop did not match` error. It also brakes the client code since I'm not getting the initial value I should be getting. And yes, I am using this custom hook inside a client component.

I've attached the link to the repository. Click to see the [custom hook](https://github.com/kaganAhmetOkan/akos-devjou/blob/main/hooks/usePersistingState.ts). And this is [the component](https://github.com/kaganAhmetOkan/akos-devjou/blob/main/components/Header/Header.tsx) that uses the custom hook.

### Additional information

```js
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Home
Binaries:
  Node: 20.10.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.1.0
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Example

https://github.com/kaganAhmetOkan/akos-devjou

**Comment by icyJoseph** [comment]
> , calling localStorage should be perfectly fine since this is a custom hook, but Next.js runs this function from the server which causes an internal server error.

Why do you think this? As the component that calls that hook is rendered during SSR, all state hooks are initialized. A custom hook is but a façade, for lack of inlining effort. That is, you could just inline the custom hook, but since you wanna share the behavior across your app, you create a custom hook, that's all there is to it.

- https://github.com/reactwg/server-components/discussions/4

You have two options here:

- Do proper sync with LocalStorage
- Make sure the hook is only called client side

To sync means, that you start your state empty, and then immediately within a useEffect or useLayoutEffect (read upon useIsomorphicLayout), synchronize the state with the browser. 

And to make sure the hook is only called client side, means to either hide it behind a flag that's set in the client, using React, or to dynamically load the component. 

- https://demystifying-rsc.vercel.app/client-components/next-dynamic/
- https://demystifying-rsc.vercel.app/client-components/no-ssr/
